Add poisoning documentation. (#2826)
dotnet/source-build: Documentation/leak-detection.md

# Leak detection (poisoning)

## Overview

> Source-build includes a mechanism for *poisoning* its input files, and then for checking for that poison after the build is complete.  This allows us to ensure that all output files were built during the build rather than copied directly from the input, which would be an illegal prebuilt usage.

Reviewer: ```suggestion
Source-build includes a mechanism for *poisoning* its input files, and then for checking for that poison int the build output.  This allows us to ensure that all output files were built during the build rather than copied directly from the input, which would be an illegal prebuilt usage.
```

Author: Changed as suggested.

## Before the build

Before the build, the MarkAndCatalogFiles task runs.  This does a few things:

> - Record the hash of every file in the source-build binary input directories (prebuilt and previously-source-built.  future expansion: reference-packages).  If the file is a zip, tarball, or nupkg, unpack it and do the same thing recursively.

Reviewer: ```suggestion
- Record the hash of every file in the source-build binary input directories (prebuilts and previously-source-built.  future expansion: [reference-packages](https://github.com/dotnet/source-build/issues/2817)).  If the file is a zip, tarball, or nupkg, unpack it and do the same thing recursively.
```

> - Record the hash of every file in the source-build binary input directories (prebuilts and previously-source-built.  future expansion: [reference-packages](https://github.com/dotnet/source-build/issues/2817)).  If the file is a zip, tarball, or nupkg, unpack it and do the same thing recursively.

Reviewer: ```suggestion
- Record the hash of every file in the source-build binary input directories (prebuilts and previously-source-built, in the future [reference-packages](https://github.com/dotnet/source-build/issues/2817)).  If the file is a zip, tarball, or nupkg, unpack it and do the same thing recursively.
```

- For managed DLLs, either bare or in an archive, add a custom attribute that marks the file as poisoned.
- For nupkgs, drop a `.poisoned` file.

> - Repack the poisoned assemblies and extra files, removing signatures so they don't fail verification.

Reviewer: Is "removing signatures" specific to the NuGet packages or does this apply to assemblies as well?

Author: Specified that this is the nupkgs only.

- Replace the binary inputs with these new packages and archives.
- Record the hash of each file again, to make sure we will be able to track output files whether they were used poisoned or unpoisoned.

## During the build

There's no change in source-build operation in poisoning mode during the build.

## After the build

> After the build, the CheckForPoison task is run on the source-build output directory.  It again unpacks any archives and packages recursively, and checks for all three kinds of markers injected before the build.  It then writes out a [report](poison-report-format.md) that details everything that was found.  This report is compared to the poison baseline to determine if the build should be marked with a warning or not. 

Reviewer: ```suggestion
After the build, the CheckForPoison task is run on the source-build output directory.  It again unpacks any archives and packages recursively, and checks for all three kinds of markers injected before the build.  It then writes out a [report](poison-report-format.md) that details everything that was found.
```

I would shy away from the errors as warnings as that is a point in time decision.

This document should mention how to enable poisoning and that there is a baseline test.
